src/main/java/com/google/firebase/auth/ListTenantsPage.java

```diff
@@ -0,0 +1,273 @@
+
+    TenantIterable(@NonNull ListTenantsPage startingPage) {
+      this.startingPage = checkNotNull(startingPage, "starting page must not be null");
+    }
+
+    @Override
+    @NonNull
+    public Iterator<Tenant> iterator() {
+      return new TenantIterator(startingPage);
+    }
+
+    /**
+     * An {@link Iterator} that cycles through tenants, one at a time.
+     * 
+     * <p>It buffers the last retrieved batch of tenants in memory. The {@code maxResults} parameter
+     * is an upper bound on the batch size.
+     */
+    private static class TenantIterator implements Iterator<Tenant> {
+
+      private ListTenantsPage currentPage;
+      private List<Tenant> batch;
+      private int index = 0;
+
+      private TenantIterator(ListTenantsPage startingPage) {
+        setCurrentPage(startingPage);
+      }
+
+      @Override
+      public boolean hasNext() {
+        if (index == batch.size()) {
+          if (currentPage.hasNextPage()) {
+            setCurrentPage(currentPage.getNextPage());
+          } else {
+            return false;
+          }
+        }
+
+        return index < batch.size();
+      }
+
+      @Override
+      public Tenant next() {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+        return batch.get(index++);
+      }
+
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException("remove operation not supported");
+      }
+
+      private void setCurrentPage(ListTenantsPage page) {
+        this.currentPage = checkNotNull(page);
+        this.batch = ImmutableList.copyOf(page.getValues());
+        this.index = 0;
+      }
+    }
+  }
+
+  /**
+   * Represents a source of tenant data that can be queried to load a batch of tenants.
+   */
+  interface TenantSource {
+    @NonNull
+    ListTenantsResult fetch(int maxResults, String pageToken) throws FirebaseAuthException;
+  }
+
+  static class DefaultTenantSource implements TenantSource {
+
+    private final FirebaseUserManager userManager;
+    private final JsonFactory jsonFactory;
+
+    DefaultTenantSource(FirebaseUserManager userManager, JsonFactory jsonFactory) {
+      this.userManager = checkNotNull(userManager, "user manager must not be null");
+      this.jsonFactory = checkNotNull(jsonFactory, "json factory must not be null");
+    }
+
+    @Override
+    public ListTenantsResult fetch(int maxResults, String pageToken) throws FirebaseAuthException {
+      ListTenantsResponse response = userManager.listTenants(maxResults, pageToken);
+      ImmutableList.Builder<Tenant> builder = ImmutableList.builder();
+      if (response.hasTenants()) {
+        builder.addAll(response.getTenants());
+      }
+      String nextPageToken = response.getPageToken() != null
+          ? response.getPageToken() : END_OF_LIST;
+      return new ListTenantsResult(builder.build(), nextPageToken);
+    }
+  }
+
+  static final class ListTenantsResult {
```

Comment:
You probably don't need this type. Just try to use the ListTenantsResponse type directly.

`listUsers()` API use this middle-ground type since the REST API returns a `User` type which needs to be mapped to `ExportedUserRecord`. But that requirement is not present here.

Reply:
Okay, but I had to add a constructor to ListTenantsResponse so that I could construct it in the unit test.